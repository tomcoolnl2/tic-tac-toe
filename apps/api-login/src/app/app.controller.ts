import {
	Controller,
	Body,
	Get,
	Post,
	HttpException,
	HttpStatus,
} from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { User, PlayerSymbol } from '@tic-tac-toe/model';

@Controller()
export class AppController {
	constructor(private readonly configService: ConfigService) {}

	@Get('username')
	getUserName(): Partial<User> {
		const name: string = this.configService.get('VITE_USERNAME');
		if (name) {
			return { name };
		} else {
			throw new HttpException(
				`Couldn't find username`,
				HttpStatus.NOT_FOUND
			);
		}
	}

	@Post('login')
	login(@Body('pwd') pwd: string): User {
		const validPwd = this.configService.get('VITE_USERPWD');
		if (pwd === validPwd) {
			return {
				name: this.configService.get('VITE_USERNAME'),
				avatar: PlayerSymbol.X,
				loggedIn: true,
			};
		} else {
			throw new HttpException(
				`Validation failed: wrong password!`,
				HttpStatus.FORBIDDEN
			);
		}
	}
}
