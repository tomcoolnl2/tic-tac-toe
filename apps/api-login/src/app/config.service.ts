// config.service.ts

import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

@Injectable()
export class EnvConfigService {
	constructor(private readonly configService: ConfigService) {}

	get(key: string): string | undefined {
		return this.configService.get<string>(key);
	}
}

@Injectable()
export class EnvConfigProvider {
	constructor(private readonly envConfigService: EnvConfigService) {}

	provideConfig(): EnvConfigService {
		return this.envConfigService;
	}
}
