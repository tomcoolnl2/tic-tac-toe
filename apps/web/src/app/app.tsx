import classNames from 'classnames';
import React from 'react';
import * as Rx from 'rxjs';
import { signIn, signOut, getCurrentUser } from 'aws-amplify/auth';
import { AppStore } from '@tic-tac-toe/core';
import { isDevEnvironment } from '@tic-tac-toe/debug';
import * as TTTModel from '@tic-tac-toe/model';
import * as TTTUI from '@tic-tac-toe/ui';

export const App: React.FC = () => {
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	const theme = React.useMemo(() => (window as any)?.electron?.theme ?? 'web', []);

	const [isSignedIn, setIsSignedIn] = React.useState<boolean>(false);
	const [authError, setAuthError] = React.useState<Error | null>(null);

	const { useContentContext } = TTTUI.Context;
	const { useBehaviorSubjectState, useScreenOrientation, useGameHandlers, useUIHandlers } =
		TTTUI.Hooks;

	const { appContent, isContentLoading, setLanguage } = useContentContext();
	const [appState] = useBehaviorSubjectState<TTTModel.AppState>(AppStore.state$);
	const { handleQuitGame, handleNextRound } = useGameHandlers(appState);
	const { openModalScreen, closeModalScreen, validateCloseModal } = useUIHandlers(appState);

	const orientation = useScreenOrientation();
	const useLandscapeDesign = React.useMemo(() => {
		return orientation?.startsWith('landscape') && 'ontouchstart' in window;
	}, [orientation]);

	React.useEffect(() => {
		(async () => {
			try {
				const currentUser = await getCurrentUser();
				console.log(`currentUser: `, currentUser);
				setIsSignedIn(true);
			} catch (err) {
				setIsSignedIn(false);
			}
		})();
	}, []);

	React.useEffect(() => {
		if (isDevEnvironment()) {
			console.info('appContent', appContent);
		}
		if (appContent) {
			document.title = appContent.appTitle;
		}
		if (appState.language) {
			setLanguage(appState.language);
		}
	}, [appContent, appState.language, setLanguage]);

	React.useEffect(() => {
		if (isDevEnvironment()) {
			console.info('appState', appState);
		}
	}, [appState, setLanguage]);

	React.useEffect(() => {
		let appScreen: TTTModel.AppScreen;
		if (isContentLoading) {
			appScreen = TTTModel.AppScreen.LOADING;
		} else if (!isSignedIn) {
			appScreen = TTTModel.AppScreen.LOGIN;
		} else {
			appScreen = TTTModel.AppScreen.SETTINGS;
		}
		AppStore.nextState({
			...AppStore.initialState,
			appScreen,
			language: appState.language,
			gameState: TTTModel.GameState.PREPLAY,
		});
	}, [isContentLoading, isSignedIn, appState.language]);

	const handleSignIn = React.useCallback(async (username: string, password: string) => {
		try {
			const { isSignedIn } = await signIn({ username, password });
			if (isSignedIn) {
				setIsSignedIn(isSignedIn);
			} else {
				throw new Error('Error signing in');
			}
		} catch (error: unknown) {
			const authError = error instanceof Error ? error : new Error('');
			setAuthError(authError);
			setIsSignedIn(false);
		}
	}, []);

	const handleSignOut = React.useCallback(async () => {
		try {
			await signOut();
		} catch (error) {
			const authError = error instanceof Error ? error : new Error('Error signing out');
			setAuthError(authError);
			console.log('Error signing out: ', error);
		}
		setIsSignedIn(false);
	}, []);

	React.useEffect(() => {
		const keyDownHandler = Rx.fromEvent<KeyboardEvent>(document, 'keydown').pipe(
			Rx.filter((event) => event.key === 'Escape')
		);
		const subscription = keyDownHandler.subscribe(validateCloseModal);
		return () => {
			subscription.unsubscribe();
		};
	}, [validateCloseModal]);

	return (
		<TTTUI.Theme theme={theme}>
			<div className={`screen ${appState.appScreen}`}>
				<div className={classNames('screen-inner', { landscape: useLandscapeDesign })}>
					<TTTUI.Error.ErrorBoundary fallback={<TTTUI.ErrorScreen />}>
						<div className="screen-front">
							{appState.appScreen === TTTModel.AppScreen.LOADING && (
								<TTTUI.LoadingScreen />
							)}
							{appContent && appState.appScreen === TTTModel.AppScreen.LOGIN && (
								<TTTUI.LoginScreen
									content={appContent.loginScreen}
									authError={authError}
									handleSubmit={handleSignIn}
									setAuthError={setAuthError}
								/>
							)}
							{appContent && appState.appScreen === TTTModel.AppScreen.SETTINGS && (
								<TTTUI.SettingsScreen
									content={appContent.settingsScreen}
									playerSymbol={appState.playerSymbol}
									selectedDifficultySetting={appState.intelligenceLevel}
									handleLogout={handleSignOut}
								/>
							)}
						</div>
						<div className="screen-back">
							{appContent && (
								<TTTUI.GameScreen
									content={appContent.gameScreen}
									pauseGame={() =>
										openModalScreen(TTTModel.AppModalScreen.PAUSED)
									}
									openRestartModal={() =>
										openModalScreen(TTTModel.AppModalScreen.RESTART)
									}
									useLandscapeDesign={useLandscapeDesign}
								/>
							)}
						</div>
					</TTTUI.Error.ErrorBoundary>
				</div>
			</div>
			{appContent && appState.appModalScreen !== null && (
				<TTTUI.Modal>
					{appState.appModalScreen === TTTModel.AppModalScreen.PAUSED && (
						<TTTUI.ResumeModalScreen
							content={appContent.modalResumeGame}
							handleResumeGame={closeModalScreen}
						/>
					)}
					{appState.appModalScreen === TTTModel.AppModalScreen.RESTART && (
						<TTTUI.RestartModalScreen
							content={appContent.restartModal}
							handleQuitGame={handleQuitGame}
							closeModalScreen={closeModalScreen}
						/>
					)}
					{appState.appModalScreen === TTTModel.AppModalScreen.GAME_OVER && (
						<TTTUI.GameOverModalScreen
							content={appContent.gameOverModal}
							gameState={appState.gameState}
							playerSymbol={appState.playerSymbol}
							cpuSymbol={appState.cpuSymbol}
							handleQuitGame={handleQuitGame}
							handleNextRound={handleNextRound}
						/>
					)}
				</TTTUI.Modal>
			)}
		</TTTUI.Theme>
	);
};
