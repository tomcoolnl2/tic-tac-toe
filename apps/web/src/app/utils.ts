export const getEventTargetElement = (event: MouseEvent): HTMLButtonElement => {
    return event.target as HTMLButtonElement;
}

export const getDataSetAttribute =
    (attribute: string) => (element: HTMLButtonElement) =>
        element.dataset[attribute];


export const delay = (ms = 500): Promise<void> => {
    return new Promise((resolve) => setTimeout(resolve, ms));
}