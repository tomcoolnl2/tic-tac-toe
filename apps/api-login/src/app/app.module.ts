import { join } from 'path';
import { Module } from '@nestjs/common';
import { ConfigModule } from '@nestjs/config';
import { AppController } from './app.controller';
import { EnvConfigProvider, EnvConfigService } from './config.service';

@Module({
	imports: [
		ConfigModule.forRoot({
			isGlobal: true,
			envFilePath: join(__dirname, '../../', '.env'),
		}),
	],
	controllers: [AppController],
	providers: [EnvConfigService, EnvConfigProvider],
	exports: [EnvConfigProvider],
})
export class AppModule {}
