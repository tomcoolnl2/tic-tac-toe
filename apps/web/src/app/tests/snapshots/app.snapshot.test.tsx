
import { render } from '@testing-library/react'
import { App } from '../../app'

describe('App component snapshot test', () => {
    it('should match the snapshot', () => {
        const { asFragment } = render(<App />)
        expect(asFragment()).toMatchSnapshot()
    })
})