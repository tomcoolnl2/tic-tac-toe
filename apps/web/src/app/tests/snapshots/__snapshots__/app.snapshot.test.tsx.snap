// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component snapshot test should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="screen settings"
  >
    <div
      class="screen-inner "
    >
      <div
        class="screen-front"
      >
        <div
          class="flex-box direction-column justify-content-flex-start align-items-center"
        >
          <div
            class="flex-box direction-row justify-content-flex-start align-items-flex-start spacing-s logo"
          >
            <span
              class="avatar avatar-x avatar-size-m"
            />
            <span
              class="avatar avatar-o avatar-size-m"
            />
          </div>
          <hr
            class="divider full-width horizontal margin-vertical-l invisible"
          />
          <span>
            Pic Player 1's Mark
          </span>
          <label
            class="choose-player"
            for="choose-player"
          >
            <input
              id="choose-player"
              type="checkbox"
            />
            <span
              class="choose-player-toggle"
            />
            <div
              class="container cols-2 choose-player-symbols"
              style="--row-height: 1fr; --col-width: 1fr;"
            >
              <div
                style="place-self: center;"
              >
                <span
                  class="avatar avatar-x avatar-size-m avatar-variant-dark"
                />
              </div>
              <div
                style="place-self: center;"
              >
                <span
                  class="avatar avatar-o avatar-size-m avatar-variant-light"
                />
              </div>
            </div>
          </label>
          <sub>
            Remember, X goes first
          </sub>
          <hr
            class="divider full-width horizontal margin-vertical-l invisible"
          />
          <div
            class="flex-box direction-column justify-content-flex-start align-items-center choose-difficulty"
          >
            <div
              class="flex-box direction-row justify-content-flex-start align-items-flex-start"
            >
              <input
                id="master"
                name="level"
                type="radio"
                value="master"
              />
              <label
                for="master"
                title="master"
              />
              <input
                id="novice"
                name="level"
                type="radio"
                value="novice"
              />
              <label
                for="novice"
                title="novice"
              />
              <input
                checked=""
                id="bieber"
                name="level"
                type="radio"
                value="bieber"
              />
              <label
                for="bieber"
                title="bieber"
              />
            </div>
            <hr
              class="divider full-width horizontal margin-vertical-s invisible"
            />
            <sup>
              bieber
            </sup>
          </div>
          <hr
            class="divider full-width horizontal margin-vertical-l invisible"
          />
          <button
            class="button button-secondary"
          >
            Start game 
            <i
              class="icon-play-sign"
            />
          </button>
        </div>
      </div>
      <div
        class="screen-back"
      >
        <div
          class="container cols-1 col-gap-m row-gap-l"
          style="--row-height: 1fr; --col-width: 1fr;"
        >
          <div>
            <div
              class="container col-gap-m row-gap-l"
              style="--row-height: 1fr; --col-width: 1fr; grid-template-columns: 1fr auto 1fr;"
            >
              <div>
                <div
                  class="flex-box direction-row justify-content-flex-start align-items-flex-start spacing-s logo"
                >
                  <span
                    class="avatar avatar-x avatar-size-m"
                  />
                  <span
                    class="avatar avatar-o avatar-size-m"
                  />
                </div>
              </div>
              <div>
                <button
                  class="button button-dark turn-indicator turn-indicator-x"
                  disabled=""
                >
                  <span
                    class="avatar avatar-x avatar-size-s avatar-variant-light"
                  />
                  Your Turn
                </button>
              </div>
              <div
                style="place-self: flex-end;"
              >
                <button
                  class="button button-light icon-reload"
                >
                  <i
                    class="icon-repeat"
                  />
                </button>
              </div>
            </div>
          </div>
          <div>
            <div
              class="container cols-3 col-gap-m row-gap-m board"
              style="--row-height: 1fr; --col-width: 1fr;"
            >
              <button
                class="cell "
                data-index="0"
                data-testid="cell-0"
                value="0"
              />
              <button
                class="cell "
                data-index="1"
                data-testid="cell-1"
                value="1"
              />
              <button
                class="cell "
                data-index="2"
                data-testid="cell-2"
                value="2"
              />
              <button
                class="cell "
                data-index="3"
                data-testid="cell-3"
                value="3"
              />
              <button
                class="cell "
                data-index="4"
                data-testid="cell-4"
                value="4"
              />
              <button
                class="cell "
                data-index="5"
                data-testid="cell-5"
                value="5"
              />
              <button
                class="cell "
                data-index="6"
                data-testid="cell-6"
                value="6"
              />
              <button
                class="cell "
                data-index="7"
                data-testid="cell-7"
                value="7"
              />
              <button
                class="cell "
                data-index="8"
                data-testid="cell-8"
                value="8"
              />
            </div>
          </div>
          <div>
            <div
              class="container cols-3 col-gap-m row-gap-m"
              style="--row-height: 1fr; --col-width: 1fr;"
            >
              <div
                class="container row-gap-s score-board-item score-board-item-x"
                style="--row-height: 1fr; --col-width: 1fr;"
              >
                <small>
                  X (You)
                </small>
                <h3>
                  0
                </h3>
              </div>
              <div
                class="container row-gap-s score-board-item "
                style="--row-height: 1fr; --col-width: 1fr;"
              >
                <small>
                  ties
                </small>
                <h3>
                  0
                </h3>
              </div>
              <div
                class="container row-gap-s score-board-item score-board-item-o"
                style="--row-height: 1fr; --col-width: 1fr;"
              >
                <small>
                  O (cpu)
                </small>
                <h3>
                  0
                </h3>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
